Even Pods Spread - 2. Calculating Predicates Metadata

Huang-Wei opened the pull request:
**What type of PR is this?**

/sig scheduling
/hold

/assign @bsalamat 
/cc @krmayankk 

**What this PR does / why we need it**:

This is the **2nd** PR of the "Even Pods Spread" KEP  implementation.

- build a new `topologyPairsPodSpreadMap` into PredicateMetadata
- update ShallowCopy()
- unit tests
- a new set of utils using "chained" pattern to build API objects

**Which issue(s) this PR fixes**:

Part of #77284.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

(will document all the changes in one place)

```release-note
NONE
```

Huang-Wei committed:
EvenPodsSpread: PredicateMetadata initilization

- build a new `topologyPairsPodSpreadMap` into PredicateMetadata
- update ShallowCopy()
- unit tests

Huang-Wei committed:
EvenPodsSpread: refactor topologyPairsPodSpreadMap

- update minMatchMap from []int32 to map[string]int32

Huang-Wei committed:
EvenPodsSpread: refactor "chained" utils

- move "chanined" utils to pkg/scheduler/testing/utils.go so as to be re-used by all scheduler tests

Huang-Wei committed:
[eps-pred-meta] auto-gen files

Huang-Wei commented:
/kind feature
/priority important-soon

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -791,3 +825,409 @@ func TestGetTPMapMatchingIncomingAffinityAntiAffinity(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.want.topologyPairToPods = make(map[topologyPair]podSet)
+			for pair, indice := range tt.injectPodPointers {
```
s/indice/indices

Huang-Wei replied:
Let's just use `indexes` to avoid confusion.

ahg-g commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -147,3 +148,193 @@ func isCSIMigrationOn(csiNode *storagev1beta1.CSINode, pluginName string) bool {
 	return mpaSet.Has(pluginName)
 }
+
+// utilities for building pod/node objects using a "chained" manner
```
I am not sure what "chained" means here, can you give an example?

ahg-g replied:
Also, where are those wrappers being used? I couldn't find any reference to them in your PRs (I think).

Huang-Wei replied:
Maybe "pipelined" is a better name? It in particular means a pattern that after calling foo(), the caller passed in is still returned so that we can streamline multiple callings with less code, like obj.foo().bar().baz()... In Kubernetes, we have similar pattern like clisentset().Namespace().Name()....

The wrappers are now only used in this feature's test cases. But I believe it can be reused to refactor other test cases.

ahg-g replied:
> The wrappers are now only used in this feature's test cases. But I believe it can be reused to refactor other test cases.

If they are to aid testing, can we move them to test-only util files? Also, it seems that the logic here is the same as the one in  "pkg/scheduler/util/utils.go" (one is exported, the other is not), can we please remove this redundancy?

Huang-Wei replied:
> can we please remove this redundancy?

Oh my bad... Apologize for the confusion. They should reside in pkg/scheduler/util/utils.go and be public.

(I guess it's due to my sloppy rebase)

> If they are to aid testing, can we move them to test-only util files?

How about creating a pkg/scheduler/util/testing.go?

ahg-g replied:
Does it need to be in a generic directory like pkg/scheduler/util? for example we have a test_util.go file under pkg/scheduler/algorithm/priorities/test_util.go, it might be better to do the same for this: place it in a file named test_util.go under the package path where it is used.

Huang-Wei replied:
For most cases that pattern works. But in this (particular) case, the helper functions are used by multiple packages<sup>1</sup>, so pkg/scheduler/util seems to be a good place. (also can avoid cyclic dependency issue)

<sup>1</sup> Precisely speaking, it's only used in one package, but will be used in generic_scheduler as well as priority package.

ahg-g replied:
ok, in that case, I think the best place for it is under the scheduler/testing, name the file utils.go and the package name is testing, how does that sound?

Otherwise, this looks good to me, and this one is just blocked on the API PR.

Huang-Wei replied:
Sounds good. Updated.

ahg-g commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -147,3 +148,193 @@ func isCSIMigrationOn(csiNode *storagev1beta1.CSINode, pluginName string) bool {
+	return p
+}
+
+// represents HARD node affinity in particular
```
comment is cut short, also please end all your comments with a period.

Huang-Wei replied:
Done.

ahg-g commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -147,3 +148,193 @@ func isCSIMigrationOn(csiNode *storagev1beta1.CSINode, pluginName string) bool {
+}
+
+// NOTE: each time we append a selectorTerm into `s`
+// and overall all selecterTerms are ORed
```
you can just say "appended selectorTerms are ORed"

Huang-Wei replied:
Done.

ahg-g commented on pkg/scheduler/util/utils.go:
```diff
@@ -140,3 +140,227 @@ func MoreImportantPod(pod1, pod2 interface{}) bool {
 	return GetPodStartTime(pod1.(*v1.Pod)).Before(GetPodStartTime(pod2.(*v1.Pod)))
 }
+
+// Utilities below to build API objects using a "chained" manner
```
ditto, where are those wrappers being used? I couldn't find any reference to them in your PRs (I think).

Huang-Wei replied:
Answered in https://github.com/kubernetes/kubernetes/pull/77760#discussion_r300890997

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -34,6 +35,9 @@ import (
 )
 
+// MaxInt32 is the maximum value of int32
+const MaxInt32 = math.MaxInt32
```
Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -91,6 +110,9 @@ type predicateMetadata struct {
 	// which should be accounted only by the extenders. This set is synthesized
 	// from scheduler extender configuration and does not change per pod.
 	ignoredExtendedResources sets.String
+	// Similar like map for pod (anti-)affinity, but impose additional min matches info
```
Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -66,6 +70,21 @@ type topologyPairsMaps struct {
+// (1) how existing pods match incoming pod on its spread constraints
+// (2) minimum match number of each hard spread constraint
+type topologyPairsPodSpreadMap struct {
+	minMatchMap map[string]int32
```
How about `tpKeyToMinMatchMap`, or `topologyToMinMatchMap`, or `minMatchedPerTopology`?

BTW: a comment has been added:

```
// This map is keyed with a topology key, and valued with the minimum number
// of pods matched on that topology domain.
```

krmayankk replied:
topologyKeyToMinPodsMap ?

ahg-g replied:
+1 to krmayankk@ suggestion.

Huang-Wei replied:
SGTM. Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -66,6 +70,21 @@ type topologyPairsMaps struct {
+// topologyPairsPodSpreadMap combines minMatchMap and topologyPairsMaps to represent
+// (1) how existing pods match incoming pod on its spread constraints
+// (2) minimum match number of each hard spread constraint
+type topologyPairsPodSpreadMap struct {
```
At this moment, I'd prefer to keep it as a standalone data structure because:

1. From the name, we can easily know it's related and only related with `EvenPodsSpread` feature.
2. Keep it standalone so as to make the code more testable.

ahg-g replied:
SG

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -66,6 +70,21 @@ type topologyPairsMaps struct {
+	*topologyPairsMaps
+}
+
+func newTopologyPairsPodSpreadMap() *topologyPairsPodSpreadMap {
```
Removed.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	}
+
+	var lock sync.Mutex
+	var firstError error
```
Good catch! I should have returned the error in this function. I will refactor this part along with your proposal about using a channel.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+			firstError = err
+		}
+		lock.Unlock()
+	}
```
Sure. Let's get to a consensus on 79777 first and I will get back to this one.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+}
+
+func podMatchesAllSpreadConstraints(pod *v1.Pod, ns string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 || pod.Namespace != ns {
```
`ns` comes from the namespace of the incoming pod.

As `constraints` has been pre-processed to only carry HARD constraints, I don't want to pass in incomingPod and re-parse, so pass in `ns` and `constraints` here - they're both from the incoming pod.

Huang-Wei replied:
There was a discussion in API review. And we will cover this limitation in the documentation.

alculquicondor replied:
sg

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+			catchError(fmt.Errorf("node %q not found", node.Name))
+			return
+		}
+		// check if current node satisfies nodeSelector / hard nodeAffinity
```
Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		for _, existingPod := range nodeInfo.Pods() {
+			ok, err := podMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
+			if err != nil {
+				catchError(err)
```
Ditto as https://github.com/kubernetes/kubernetes/pull/77760#discussion_r300890950

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -165,46 +191,217 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 
-func (topologyPairsMaps *topologyPairsMaps) removePod(deletedPod *v1.Pod) {
+// add a topology pair holder if needed
+func (m *topologyPairsMaps) addHeadlessTopologyPair(pair topologyPair) {
```
Done.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -791,3 +825,409 @@ func TestGetTPMapMatchingIncomingAffinityAntiAffinity(t *testing.T) {
+			want: true,
+		},
+		{
+			name:      "normal mis-match",
```
Done.

Huang-Wei commented:
@ahg-g Thanks a lot for the review!! Most of the comments have been addressed. PTAL.

ahg-g commented:
I think we can get rid of the topologyPairToPods map. This map is very expensive, it is in the order of number of pods, which can be huge.

We don't really need to keep track of which exact pods in the map, all we need is the count, so the map should be topologyPairToNumOfPodsMap. Add and Remove pod basically update the count, rather than push to a list.

ahg-g commented:
> I think we can get rid of the topologyPairToPods map. This map is very expensive, it is in the order of number of pods, which can be huge.
> 
> We don't really need to keep track of which exact pods in the map, all we need is the count, so the map should be topologyPairToNumOfPodsMap. Add and Remove pod basically update the count, rather than push to a list.

Note that I think this is an optimization that we have to do in this PR, but shouldn't be a blocker to the API change.

Huang-Wei commented:
I felt the same in the beginning of implementation. But in terms of preemption, in particular the removePod() logic, using a {tpPair:count} map instead of {tpPair:podSet} has the following downsides:

1. Signature of removePod() needs to be updated to include both `NodeInfo` and `Preemptor` pod:
	1. only with the `NodeInfo`, we are able to get the node labels matching the "deletedPod", so add/minus the count number.
	1. and also we need the `Preemptor` pod to check whether the "deletedPod" matches `Preemptor` pod's topologySpreadConstraints, again.
1. Using a {tpPair:count} map will result in re-calculation in every deletePod() cycle; whereas in the {tpPair:podSet} solution the calculation result is "cached".

Regarding # 1, I'm not sure if we should change the method signature to adapt to EvenPodsSpread case **_only_**.

Regarding # 2, it's essentially a classical space vs. time problem - you can't get both at the same time. If we consider preemption as a high-churn case (as a lot of pods need to be deleted in a dryrun manner), then I'd more prefer the {tpPair:podSet} solution.

@ahg-g, thoughts?

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	return copy
+}
+
+func (podSpreadMap *topologyPairsPodSpreadMap) clone() *topologyPairsPodSpreadMap {
```
Also `m` as receiver

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	}
+	copy := &topologyPairsPodSpreadMap{
+		topologyKeyToMinPodsMap: make(map[string]int32),
+		topologyPairsMaps:       newTopologyPairsMaps(),
```
`: m.topologyPairsMaps.clone()`

alculquicondor commented on pkg/scheduler/testing/utils.go:
```diff
@@ -0,0 +1,243 @@
+	return p
+}
+
+// NodeWrapper wraps a LabelSelector inside.
```
wraps a Node

alculquicondor commented on pkg/scheduler/algorithm/predicates/metadata_test.go:
```diff
@@ -25,7 +25,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
-	schedulertesting "k8s.io/kubernetes/pkg/scheduler/testing"
+	st "k8s.io/kubernetes/pkg/scheduler/testing"
```
nit: revert this line. It makes it easier to identify where an identifier is coming from.

Huang-Wei replied:
I'm more lean towards `st` as the places where `schedulertesting`/`st` is used is already indented with several tabs, saving 14 letters each line can make the code more compact and readable.

alculquicondor replied:
I'm not convinced that compact==more-readable. But since I don't think this is [standardized](https://github.com/golang/go/wiki/CodeReviewComments#imports), I'll leave the decision to you.

alculquicondor commented on pkg/scheduler/testing/utils.go:
```diff
@@ -0,0 +1,243 @@
+)
+
+// NodeSelectorWrapper wraps a NodeSelector inside.
+type NodeSelectorWrapper struct{ v1.NodeSelector }
```
maybe Builder is a more appropriate name for all these structs. And `builders.go` as filename.

Huang-Wei replied:
Sounds good.

Huang-Wei replied:
IMO `builder` doesn't sound more descriptive than `wrapper` (please correct me if anybody thing it's much better).

But I do like the idea to rename `utils.go` to `wrappers.go`/`builders.go`.

alculquicondor replied:
A wrapper would be something that adds more fields that don't exist in the original struct. This is basically holding an object until you get it from `Obj`. Hence, it knows how to build the object.

ahg-g commented:
> I felt the same in the beginning of implementation. But in terms of preemption, in particular the removePod() logic, using a {tpPair:count} map instead of {tpPair:podSet} has the following downsides:
> 
> 1. Signature of removePod() needs to be updated to include both `NodeInfo` and `Preemptor` pod:
>    
>    1. only with the `NodeInfo`, we are able to get the node labels matching the "deletedPod", so add/minus the count number.
>    2. and also we need the `Preemptor` pod to check whether the "deletedPod" matches `Preemptor` pod's topologySpreadConstraints, again.
> 2. Using a {tpPair:count} map will result in re-calculation in every deletePod() cycle; whereas in the {tpPair:podSet} solution the calculation result is "cached".
> 
> Regarding # 1, I'm not sure if we should change the method signature to adapt to EvenPodsSpread case **_only_**.
> 
> Regarding # 2, it's essentially a classical space vs. time problem - you can't get both at the same time. If we consider preemption as a high-churn case (as a lot of pods need to be deleted in a dryrun manner), then I'd more prefer the {tpPair:podSet} solution.
> 
> @ahg-g, thoughts?

Regarding #1 I don't think changing the interface is a blocker to a potential significant saving in memory. I expect even pod spreading to be used often and we should cater for it, especially that this is not really a user facing interface.


Regarding #2, right, it is a space time tradeoff, but only in the case where the cluster is under high resource pressure, otherwise we are trading space for no time gain. I don't have statistics, but my impression is that high preemption will likely trigger cluster autoscaler to scale the cluster up, or the operator manually adding more nodes, and so I expect it to be a transient.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+			cancel()
+			return
+		}
+		// Be design if NodeAffinity or NodeSelector is defined, spreading is
```
hmm... I started initially with "If NodeAffinity..." and added "By design" per ahg-g's suggestion, let me reword to: `In accordance to design, if...`

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		if !podMatchesNodeSelectorAndAffinityTerms(pod, node) {
+			return
+		}
+		// ensure current node's labels contains all topologyKeys in 'constraints'
```
Nope. If a node doesn't have desired topology key(s), it's not an eligible node and won't be considered.

alculquicondor replied:
sg, thanks for confirming

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+// 1. podLabels = nil, constraint.LabelSelector = nil => returns false
+// 2. podLabels = nil => returns false
+// 3. constraint.LabelSelector = nil => returns false
+func podLabelsMatchesSpreadConstraints(podLabels map[string]string, constraints []v1.TopologySpreadConstraint) (bool, error) {
```
It's to check if all constraints[*].labelSelector matches podLabels. Renaming the function to `spreadConstraintsMatchPodLabels` makes more sense?

alculquicondor replied:
I'm not questioning what the function is doing, but rather why. In line 248, we are ignoring the pod if "all labels" don't match. I'm arguing that it should be "at least one label". AND vs OR.

ahg-g replied:
This is a good point Aldo. 

@Huang-Wei if two constraints have different label selectors, we should be computing the skew for each one separately, and so the min match may need to be computed per topology per "label selector", right?

Huang-Wei replied:
Oh, I understand your concerns now. This is a very good point and it's not the first time we face this issue - [Support inter-Pod affinity to one or more Pods #68701](https://github.com/kubernetes/kubernetes/issues/68701) is a pretty similar one.

All constraints are evaluated with a `AND` semantics. But even the `AND` can be ambiguous:

1. all constraints must be matched on the same pod
2. all constraints must be matched, but can match on different pods

Current implementation follows the 1st explanation, which is simpler and less error-prone. In the alpha phase, IMO it's good to start with 1st implementation which is also consistent with current PodAffinity semantics.
But I do respect endusers' opinion on this. If users have strong preference on the 2nd, we can update in the next release cycle.

cc/ @krmayankk @bsalamat

ahg-g replied:
I think option 1 is counter intuitive, and I would even argue wrong. In the following setup for example, pod 2 has no impact on the skew calculation when scheduling pod1, but pod1 impacts the scheduling of pod2 if it were to be scheduled first. 

If we want simplify the feature and go with option 1, then matchLabels should be defined outside the topologySpreadConstraint.

Pod 1 spec:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: bar
```

Pod 2 spec:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
```

alculquicondor replied:
I figured I should share the analysis I made of the performance implications of supporting option 2.

This is the current implementation in pseudocode.
```
For each Node (parallel)
  For each Pod
    Match all labels or continue
    For each constraint
      map[topologyKey, value].append(Pod)

For each (topologyKey, value) in map:
  minMap[topologyKey] = min(minMap[topologyKey, len(map[topologyKey, value]))
```

If we want to support counts per label, the algorithm (and complexity) doesn't need to change much.
We move the label match inside the constraints loop. And we keep a set of counters (and minSkew) for each constraint. It can be indexed in a slice. Something like: `[]map[string]int`, where the first index is the number of constraint, then the values of the topologyKey, and then we just have the counter.

In conclusion, the performance is the same. Note that this option also suppresses the need to validate that constraints target different topology keys (which was part of the API validation PR). I think we should do this change sooner rather than later, because it is user facing.

Huang-Wei replied:
> Note that this option also suppresses the need to validate that constraints target different topology keys (which was part of the API validation PR).

The API validation PR is different - that validates the same tuple {topologyKey, scheduleAction} must be unique.

alculquicondor replied:
> The API validation PR is different - that validates the same tuple {topologyKey, scheduleAction} must be unique.

We are saying that counters need to be per {topologyKey, action, labels}. So it is valid to have different constraints with the same {topologyKey, action}. And at that point, we can simply allow repetitions in general. Having more than one constrain with the same triple is unnecessary for a user, but it's semantically correct.

In the example from @ahg-g, having the same topologyKey for both constraints makes sense under this interpretation.

Huang-Wei replied:
> Having more than one constrain with the same triple is unnecessary for a user, but it's semantically correct.

We should give a validation error to save calculation effort in scheduler side. For the following case:

```yaml
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: bar
```

It's easier to understand with a "merged" yaml like:

```yaml
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchExpressions:
        key: app
        operator: In
        values: ["app", "bar"]
```

alculquicondor replied:
Good point. You could theoretically have different maxSkew for each label. But I don't think that's a useful use-case. Keeping the validation SGTM.

bsalamat replied:
If we have a few constraints, the label selectors of each constraint may or may not match a pod. The constraints whose labels do not match a pod are considered satisfied for that pod. Let's use @ahg-g example:

incoming pod:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/region
    selector:
      matchLabels:
        app: bar
```

existing pod:
```
labels:
   app: foo
```

The existing pod matches only the first constraint of the incoming pod. So, we count the existing pod towards the pods in its corresponding `zone`. The existing pod does not affect the count that we keep for each `region` (second constraints) when we check `maxSkew`.

ahg-g replied:
That is what I am suggesting that the behavior should be as well, but I don't think the implementation actually behaves this way. May be I am missing something, but it seems that in line [243](https://github.com/kubernetes/kubernetes/blob/d5e205af21ec42f98f0e8b41f3db105e9c566473/pkg/scheduler/algorithm/predicates/metadata.go#L243) we check if the existing pod matches all the constraints, and only then we add it to the minMatch of all the toplogyKeys included in the constraints. So in the example, the existing pod will not be even be counted towards the pods in the corresponding zone.

Huang-Wei replied:
@ahg-g You read it right on current implementation.

Ok, as a majority of people prefer option 2, I will make a commit in this PR to address this.

Thanks @bsalamat @ahg-g! and @alculquicondor for bringing this up!

Huang-Wei replied:
Logic updated as an additional commit. PTAL.

BTW: I will update downstream PRs as well.

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	}
+	podLabelSet := labels.Set(podLabels)
+	for _, constraint := range constraints {
+		selector, err := metav1.LabelSelectorAsSelector(constraint.LabelSelector)
```
It's a suggestion but not implemented yet. It's tracked in the "umbrella" issue.

alculquicondor replied:
sg

Huang-Wei committed:
fixup: fix comments and use a channel to pass err

Huang-Wei commented:
@alculquicondor Thanks for the review! Comments addressed, PTAL.

@ahg-g Points taken. But in order not to cause transitive rebasing hassling, and keep the current implementation as a record/commit (which functions well), I'd propose to keep it as is.
As a potential enhancement, I will create a PR after all PRs get merged, so that if the {tpPair:count} solution isn't working well someday, at least we can revert it back.
If we apply the {tpPair:count} implementation in place in this PR, we just lose current {tpPair:podSet} codebase and have no way to revert back :)

ahg-g commented:
> @alculquicondor Thanks for the review! Comments addressed, PTAL.
> 
> @ahg-g Points taken. But in order not to cause transitive rebasing hassling, and keep the current implementation as a record/commit (which functions well), I'd propose to keep it as is.
> As a potential enhancement, I will create a PR after all PRs get merged, so that if the {tpPair:count} solution isn't working well someday, at least we can revert it back.
> If we apply the {tpPair:count} implementation in place in this PR, we just lose current {tpPair:podSet} codebase and have no way to revert back :)

Trying to optimize it later sounds good to me.

alculquicondor commented:
> I felt the same in the beginning of implementation. But in terms of preemption, in particular the removePod() logic, using a {tpPair:count} map instead of {tpPair:podSet} has the following downsides:
> 
> 1. Signature of removePod() needs to be updated to include both `NodeInfo` and `Preemptor` pod:
>    
>    1. only with the `NodeInfo`, we are able to get the node labels matching the "deletedPod", so add/minus the count number.
>    2. and also we need the `Preemptor` pod to check whether the "deletedPod" matches `Preemptor` pod's topologySpreadConstraints, again.

I'm looking at the current removePod implementation and I don't see why we need the node. We already have `podSpreadMap.podToTopologyPairs[deletedPodFullName]` which tells us exactly which pair's counters to update. Am I missing something?

Huang-Wei commented:
> I'm looking at the current removePod implementation and I don't see why we need the pod's node. We already have `podSpreadMap.podToTopologyPairs[deletedPodFullName]` which tells us exactly which pair's counters to update. Am I missing something?

That discussion was based on the context of switching to the {tpPair:count} solution. In that case, we don't hold a {podName:tpPairSet} structure.

draveness commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,218 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 	podFullName := schedutil.GetPodFullName(pod)
-	if topologyPairsMaps.topologyPairToPods[pair] == nil {
-		topologyPairsMaps.topologyPairToPods[pair] = make(map[*v1.Pod]struct{})
+	if m.topologyPairToPods[pair] == nil {
```
nit: Could we replace this we `addTopologyPairWithoutPods`?

Huang-Wei replied:
Done.

draveness commented on pkg/scheduler/algorithm/predicates/utils.go:
```diff
@@ -19,7 +19,7 @@ package predicates
 import (
 	"strings"
 
-	v1 "k8s.io/api/core/v1"
```
Can we revert this change since there is nothing updates else in this file and it is not a restriction that we should import without `v1` prefix

draveness commented on pkg/scheduler/util/utils.go:
```diff
@@ -21,7 +21,7 @@ import (
 
 	"time"
 
-	v1 "k8s.io/api/core/v1"
```
same here

alculquicondor commented:
> That discussion was based on the context of switching to the {tpPair:count} solution. In that case, we don't hold a {podName:tpPairSet} structure.

I don't think we suggested that. Only to get rid of {tpPair:podSet}. I think we should still be able to hold {podName:tpPairSet}, which can even point to the same map (same address) for all the pods in the same node.

Huang-Wei commented:
> I don't think we suggested that. Only to get rid of {tpPair:podSet}. I think we should still be able to hold {podName:tpPairSet}, which can even point to the same map (same address) for all the pods in the same node.

That's an option. Let's see how it can benefit upon eliminating {tpPair:podSet}.

alculquicondor commented:
> > I don't think we suggested that. Only to get rid of {tpPair:podSet}. I think we should still be able to hold {podName:tpPairSet}, which can even point to the same map (same address) for all the pods in the same node.
> 
> That's an option. Let's see how it can benefit upon eliminating {tpPair:podSet}.

sgtm

Huang-Wei committed:
EvenPodsSpread: match selector of each constraint independently

- see more discussion at https://github.com/kubernetes/kubernetes/pull/77760#discussion_r305107973

Huang-Wei commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,218 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 }
 
-func (topologyPairsMaps *topologyPairsMaps) removePod(deletedPod *v1.Pod) {
+// add a topology pair holder if needed
```
It's usually tolerant for private functions.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

alculquicondor commented:
/lgtm

ahg-g commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -248,22 +240,25 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
+				}
+				if ok {
 					// constraint.TopologyKey is already guaranteed to be present
 					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
```
You may also want to add that since we restrict a toplogyKey to exist in a single hard constraint, using topolgyKey as the key in this map is good enough to identify which exact constraint this pod is counted against (because otherwise the key to the map should be some sort of an identifier that combines the topologyKey and the label selector)

ahg-g reviewed (commented):
/lgtm

Thanks for modifying it.

Huang-Wei commented:
/hold cancel

tedyu commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,206 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[constraint.TopologyKey] = math.MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
```
Is it possible that value of some topologyPairsPodSpreadMap.topologyKeyToMinPodsMap is
math.MaxInt32 at the end of this loop ?

Huang-Wei replied:
Nope. The logic below ensures for each tp pair, they're valued with either empty podSet, or a concrete podSet:

https://github.com/kubernetes/kubernetes/blob/5993ec5a5a306de2b329ae7bfd182e8dddf3a8f3/pkg/scheduler/algorithm/predicates/metadata.go#L260-L272

Then in the above loop, the default value (MaxInt32) will be compared with len(podSet), so they won't remain MaxInt32 at the end of the loop.

tedyu commented on pkg/scheduler/algorithm/predicates/metadata.go:
```diff
@@ -166,46 +189,206 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[constraint.TopologyKey] = math.MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit unvisited portions of <topologyKey: any value>
```
I wonder what type of short circuiting you were thinking about.
Without short circuiting, the following comparison is performed and we continue to the next pair:
```
l < topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key]
```
